### DISCUSSION TITLE
Malloc failing when called from firmware

### DISCUSSION THREAD

**Discussion by Larpncohez** [micropython/micropython#15882] [RP2040 / Pico]
I started by using the guide [here](https://www.hackster.io/sandeep-mistry/create-a-usb-microphone-with-the-raspberry-pi-pico-cc9bd5) to make the Adafruit PDM MEMS breakout function. I would have used the relevant circuitpython library but, from what I have seen, it isn't really meant to work with the pico. If I am wrong about that, it'd be immensely helpful.

GitHub page for this guide is [here](https://github.com/sandeepmistry/pico-microphone.git)

After getting the library to compile and function properly, I attempted to make the library accessible through micropython, deciding to opt for rewriting the firmware after struggling with MPY files.

To do this, I used the following forum thread as guidance, replacing the mentioned core1_code with a slight variant of the code in examples/hello_pdm_microphone in the aforementioned GitHub page: [forum](https://forums.raspberrypi.com/viewtopic.php?p=1948172&sid=3f32f0507698b53e5f74f11ec2428717#p1948172)

After several changes to CMakeLists and a weird series of file transfers, the module imports correctly, and after *a lot* of debugging I determined a line that seemingly causes the entire MicroPython backend to fail with Thonny being unable to recover the backend unless the pico is fully restarted. 

Found on line 50 of src/pdm_microphone.c of the GitHub:

`pdm_mic.raw_buffer[i] = malloc(pdm_mic.raw_buffer_size);`

Is malloc just not supported when called from the firmware?

**Comment by Larpncohez** [comment]
In all honesty I am likely just going to order some analog microphones and wire things more directly. This is.... painful, but it would be nice to have clarification in the future

**Comment by bidrohini** [comment]
You can try the Adafruit CircuitPython library for PDM microphones on the Pico.
